### PR TITLE
Update Responses to use 0.22.0

### PR DESCRIPTION
**Pull request by vilit1**
Update to pin responses to the newest 0.22.0. Only tests were updated.

The biggest change is that now 500 return a different error - msrest.exceptions.ClientRequestError. This is similar to what the service returns for a hub with a non working dataplane.

Query has also been changed: if you have a 500 response with a next link, the responses package will go to the next page rather than failing right away. Tests have been updated to reflect this.

Note that 0.21.0 is skipped because the 500's return urllib3.exceptions.MaxRetryError

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py [hunk 1]
```diff
@@ -420,6 +433,12 @@ def test_update_twin_error(self, fixture_cmd, service_client_error):
                 resource_group_name=None,
                 etag=None
             )
+        status_codes = [call.response.status_code for call in service_client_error.calls]
+        # Retries 5 times
```
Comment (digimaun) [inline azext_iot/tests/digitaltwins/test_dt_twin_unit.py:437]: What about having the service_client_error fixture patch the service client so it doesn't retry?

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py [hunk 2]
```diff
@@ -168,6 +169,10 @@ def test_query_twins_error(self, fixture_cmd, service_client_error):
                 show_cost=False,
                 resource_group_name=None
             )
+            # For some reason the new responses pkg doesnt fail on the first 500...
+            status_codes = [call.response.status_code for call in service_client_error.calls]
+            assert status_codes == [500, 200]
+            raise CLIError()
```
Comment (digimaun) [inline azext_iot/tests/digitaltwins/test_dt_twin_unit.py:175]: Does the service_client_error fixture need to be constructed in a different way with the new version? It is suspicious to have to raise CLIError in the test to fulfill what the application logic is supposed to throw.

File: azext_iot/tests/iothub/configurations/test_iot_config_unit.py [hunk 3]
```diff
@@ -902,10 +903,14 @@ def test_config_list_invalid_args(self, fixture_cmd, top):
 
     def test_config_list_error(self, fixture_cmd, service_client_generic_errors):
         service_client_generic_errors.assert_all_requests_are_fired = False
-        with pytest.raises(CLIError):
+        with pytest.raises((CLIError, ClientRequestError)) as e:
```
Comment (digimaun) [inline azext_iot/tests/iothub/configurations/test_iot_config_unit.py:906]: The error tests need to assert what the application logic will throw rather than the service client unless the command logic does not catch the error, why would a new responses version change that?

Comment (vilit1) [reply azext_iot/tests/iothub/configurations/test_iot_config_unit.py:906]: The ClientRequestError is not caught by the command logic and this seems to align with my testing with a broken IoT Hub.

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py (hunk 2 above)
Comment (vilit1) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:175]: This is for the test case that has the first request as a 500 (with a next link) and the second request as a 200 (from the next link). The result is that the command succeeds even though there is a 500.

I cannot test this in a real service...

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py (hunk 1 above)
Comment (vilit1) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:437]: In a real service call, a 500 ends with the retries so this is closer to a real life situation. Plus the retries in the unit test are still very fast.

Comment (digimaun) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:437]: Why is it not consistent where all other service_client_error based unit tests don't require this same setup for retries? And why does updating responses necessitate this?

Comment (vilit1) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:437]: hmm ok let me look through the other retries and update them so the tests look for the retry calls

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py (hunk 2 above)
Comment (digimaun) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:175]: Raising a CLIError in the test code that is supposed to validate the application code raises CLIError is wrong and not testing what its supposed to.

Looking at the respective digitaltwins client operation, if the response status code is not 200 it should raise models.ErrorResponseException.

image

You don't need a real service to test this flow, and at a minimum, one can change the status code of the response in a breakpoint before the response is inspected in further logic.

File: azext_iot/tests/iothub/configurations/test_iot_config_unit.py (hunk 3 above)
Comment (digimaun) [reply azext_iot/tests/iothub/configurations/test_iot_config_unit.py:906]: The client is supposed to throw a `CloudError` if the response is not 200

image

And that is what the command handler is expecting

image

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py (hunk 2 above)
Comment (vilit1) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:175]: The issue is that the code skips over the 500 (since there is a next link) and thus the final response status code is 200. So the case where the first link is a 500 with a next link to a 200 doesn't throw an error but succeeds...

Example:
`pytest azext_iot/tests/digitaltwins/test_dt_twin_unit.py::TestTwinQueryTwins::test_query_twins_error[service_client_error2]`

pdb prints:
image

Comment (vilit1) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:175]: But I do understand that raising a CLIError in this case is very weird since the code does not error. My other suggestion is to remove this test case.

Comment (digimaun) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:175]: I don't see how the code skips over the 500, when a 500 will interrupt page accumulation that happens with `accumulate_result`.

When I run the same test (with responses 0.20.0)

Right before the first API call (notice I have 2 break points here, before and after the call)

image


After getting the first response by the digitaltwins client
image

It's a 500 so raise models.ErrorResponseException ....I will continue the debugger from here to the next breakpoint
image

The next breakpoint is the one that wraps accumulate result and uses `handle_service_exception` to throw something derived from CLIError.
image

Which passes the existing test definition.

image

File: azext_iot/tests/digitaltwins/test_dt_twin_unit.py (hunk 1 above)
Comment (digimaun) [reply azext_iot/tests/digitaltwins/test_dt_twin_unit.py:437]: It's more about understanding the why this behavior in _this responses version_ before applying it everywhere.

File: azext_iot/tests/iothub/configurations/test_iot_config_unit.py (hunk 3 above)
Comment (vilit1) [reply azext_iot/tests/iothub/configurations/test_iot_config_unit.py:906]: The line above - aka the 273 response, now throws the error for a maxretry (500) service error

So the code never gets to checking the response.status_code and the error is not caught by our handle_service_exception

Comment (digimaun) [reply azext_iot/tests/iothub/configurations/test_iot_config_unit.py:906]: now implying in responses 0.22.0 where it works as expected in 0.20.0

**Comment by vilit1** [comment]
Closing this since I found a better way
